File: algo/257. Binary Tree Paths.js
// @9 Divide and Conquer

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
var binaryTreePaths = function(root) {
    
    // child of leaf
    if (root == null) {
        return [];
    }
    
    // we need to do something for leaf
    // since the resulted string is different for leaf
    // eg. "1->2->3". there is no "->" at the end 
    if (root.left == null && root.right == null) {
        return [root.val + ""];
    }
    
    // initialize parameters or utilities
    var res = [];
    
    // divide
    var leftPaths = binaryTreePaths(root.left);
    var rightPaths = binaryTreePaths(root.right);
    
    // merge
    for (var i = 0; i < leftPaths.length; i++) {
        res.push(root.val + "->" + leftPaths[i]);
    }

    for (var i = 0; i < rightPaths.length; i++) {
        res.push(root.val + "->" + rightPaths[i]);
    }
  
    return res;
};

File: round3/257. Binary Tree Paths.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<string> BinaryTreePaths(TreeNode root) {
        // exit
        if (root == null) return new List<string>();
        if (root.left == null && root.right == null) return new List<string>{$"{root.val}"}; 
        
        // divide
        var left = BinaryTreePaths(root.left);
        var right = BinaryTreePaths(root.right);
        
        // merge
        var ret = new List<string>();
        foreach (string leftPath in left) {
            ret.Add($"{root.val}->" + leftPath);
        }
        foreach (string rightPath in right) {
            ret.Add($"{root.val}->" + rightPath);
        }
        
        return ret;
    }
}